src/solvers/refinement/string_refinement.cpp

```diff
@@ -1772,9 +1777,165 @@ bool string_refinementt::is_axiom_sat(
     }
   case decision_proceduret::resultt::D_UNSATISFIABLE:
     return false;
+  case decision_proceduret::resultt::D_ERROR:
   default:
     INVARIANT(false, string_refinement_invariantt("failure in checking axiom"));
     // To tell the compiler that the previous line bails
     throw 0;
   }
 }
+
+/// \related string_constraintt
+typedef std::map<exprt, std::vector<exprt>> array_index_mapt;
+
+/// \related string_constraintt
+class gather_indices_visitort: public const_expr_visitort
+{
+public:
+  array_index_mapt indices;
+
+  gather_indices_visitort(): indices() {}
+
+  void operator()(const exprt &expr) override
+  {
+    if(expr.id()==ID_index)
+    {
+      const index_exprt index=to_index_expr(expr);
+      const exprt s(index.array());
+      const exprt i(index.index());
+      indices[s].push_back(i);
+    }
+  }
+};
+
+/// \related string_constraintt
+static array_index_mapt gather_indices(const exprt &expr)
+{
+  gather_indices_visitort v;
+  expr.visit(v);
+  return v.indices;
+}
+
+/// \related string_constraintt
+class is_linear_visitort: public const_expr_visitort
+{
+public:
+  bool correct;
+
+  is_linear_visitort(): correct(true) {}
+
+  void operator()(const exprt &expr) override
+  {
+    if(expr.id()!=ID_plus && expr.id()!=ID_minus && expr.id()!=ID_unary_minus)
+    {
+      // This represents that the expr is a valid leaf, may not be future proof
+      // or 100% enforced, but is correct prescriptively. All non-sum exprs must
+      // be leaves.
+      correct&=expr.operands().empty();
+    }
+  }
+};
+
+/// \related string_constraintt
+static bool is_linear(const exprt &expr)
+{
+  is_linear_visitort v;
+  expr.visit(v);
+  return v.correct;
+}
+
+/// \related string_constraintt
+static bool universal_only_in_index(const string_constraintt &expr)
+{
+  // We need to check that each instance of the universal variable occurs in
+  // and index expression. For efficiency, we do a depth-first search of the
+  // body. The exprt visitors do a BFS and hide the stack/queue, so we would
+  // need to store a map from child to parent.
+
+  // The int represents index depth we are.
+  typedef std::pair<exprt, unsigned> valuet;
+  std::stack<valuet> stack;
+  // How many index expressions we are currently in. We start at 0 since expr
+  // is not an index expression.
```

Comment:
These comments do not sound very clear to me. Can we rephrase and merge with the comment above?